cli-plugins/plugin/plugin.go

```diff
@@ -24,12 +32,45 @@ import (
 // called.
 var PersistentPreRunE func(*cobra.Command, []string) error
 
+// closeOnParentSocketClose connects to the socket specified
+// by the CLI_PLUGIN_EXIT_SOCKET env var, if present, and attempts
+// to read from it until it receives an EOF, which signals that
+// the CLI is going to exit and the plugin should also exit.
+func closeOnParentSocketClose() {
+	socketAddr, ok := os.LookupEnv(CliPluginCloseSocketEnvKey)
+	if !ok {
+		// if a plugin compiled against a more recent version of docker/cli
+		// is executed by an older CLI binary, ignore missing environment
+		// variable and behave as usual
+		return
+	}
+	addr, err := net.ResolveUnixAddr("unix", socketAddr)
+	if err != nil {
+		return
+	}
+	cliCloseConn, err := net.DialUnix("unix", nil, addr)
+	if err != nil {
+		return
+	}
+
+	go func() {
+		for {
+			b := make([]byte, 1)
```

Comment:
Be nice to the garbage collector; reuse your scratch buffer.
```suggestion
		b := make([]byte, 1)
		for {
```